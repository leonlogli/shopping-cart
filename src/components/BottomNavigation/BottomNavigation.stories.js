import React from 'react'
import AccountCircleOutlinedIcon from '@material-ui/icons/AccountCircleOutlined'
import RestoreIcon from '@material-ui/icons/Restore'
import FavoriteIcon from '@material-ui/icons/Favorite'

import BottomNavigation from '.'

export default {
  title: 'BottomNavigation',
}

const navLinks = [
  { label: 'Restaurants', value: '/restaurants', icon: <RestoreIcon /> },
  { label: 'Menu', value: '/menu', icon: <FavoriteIcon /> },
  { label: 'Profil', value: '/profile', icon: <AccountCircleOutlinedIcon /> },
]

export const Default = () => <BottomNavigation links={navLinks} value="/menu" />
