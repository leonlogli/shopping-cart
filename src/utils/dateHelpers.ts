const durationFormat = (interval: number, type: string) => {
  return `il y a ${Math.floor(interval)} ${type}${
    Math.floor(interval) > 1 ? 's' : ''
  }`
}

const fromNow = (date: Date) => {
  const seconds = Math.floor(
    ((new Date() as any) - (new Date(date) as any)) / 1000
  )

  let interval = seconds / 31536000

  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'an')
  }

  interval = seconds / 2592000
  if (Math.floor(interval) >= 1) {
    return `il y a ${Math.floor(interval)} mois`
  }

  interval = seconds / 86400
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'jour')
  }

  interval = seconds / 3600
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'heure')
  }

  interval = seconds / 60
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'minute')
  }

  return `Ã  l'instant`
}

/**
 * Parse the given string time to a date.Time must have `hh:mm:ss` format
 * @param time time to parse
 * @param date date to which the time will be set once parsed. If null, today's date is used
 */
const parseTimeToDate = (time: string, date?: Date) => {
  const dateToSet = date || new Date()
  const timeParts = time.split(':').map((t) => Number(t) || 0)

  dateToSet.setHours(timeParts[0], timeParts[1], timeParts[2] || 0)

  return dateToSet
}

const formatShortDate = (date: any, locale?: string) => {
  return new Date(date).toLocaleDateString(locale, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  })
}

export { fromNow, parseTimeToDate, formatShortDate }
