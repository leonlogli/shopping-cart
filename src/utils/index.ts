export * from './testHelpers'
export * from './util'
export * from './dateHelpers'
