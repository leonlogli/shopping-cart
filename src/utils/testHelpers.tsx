import React from 'react'
import { ThemeProvider } from 'styled-components'
import { mount, shallow } from 'enzyme'
import { defaultTheme } from '../styles'

const ThemeProviderWrapper = ({ children }: any) => (
  <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
)

export const shallowWithTheme = (tree: any) =>
  shallow(tree, {
    wrappingComponent: ThemeProviderWrapper,
  })

export const mountWithTheme = (tree: any) =>
  mount(tree, {
    wrappingComponent: ThemeProviderWrapper,
  })
