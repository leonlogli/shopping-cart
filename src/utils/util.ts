const toUSD = (value: number) => {
  return `${value.toLocaleString('en-US', {
    minimumFractionDigits: 2,
  })} USD`
}

export { toUSD }
