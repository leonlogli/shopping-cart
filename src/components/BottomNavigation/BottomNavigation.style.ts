import styled from 'styled-components'
import BottomNavigation from '@material-ui/core/BottomNavigation'

const BottomNavigationRoot = styled(BottomNavigation)`
  width: 100%;
  max-width: 767px;
  background: ${({ theme }) => theme.palette.primary.main};

  .MuiBottomNavigationAction-root {
    color: whitesmoke;

    &.Mui-selected {
      color: white;
    }
  }
`

export { BottomNavigationRoot }
