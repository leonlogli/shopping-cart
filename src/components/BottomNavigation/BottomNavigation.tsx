import React, { ChangeEvent, useState, useEffect } from 'react'
import BottomNavigationAction from '@material-ui/core/BottomNavigationAction'
import { BottomNavigationProps as BaseProps } from '@material-ui/core/BottomNavigation'

import { BottomNavigationRoot } from './BottomNavigation.style'

type NavLink = {
  label: string
  value: string
  icon: React.ReactNode
}

export type BottomNavigationProps = Omit<BaseProps, 'onChange'> & {
  links: NavLink[]
  whiteListValues?: string[]
  onChange?: (value: string) => void
}

/**
 * Bottom navigation component. Each nav link is automatically focus if pressed.
 * To disable focus for some nav links, just prepend `#` to their value (Ex: '#search')
 */
const BottomNavigation = (props: BottomNavigationProps) => {
  const { links, value: valueProp, onChange, ...other } = props

  const initialLink =
    links.find((link) => link.value === valueProp) ||
    links[Number(valueProp)] ||
    links[0]

  const [value, setValue] = useState(initialLink.value)

  useEffect(() => {
    setValue(valueProp)
  }, [valueProp])

  const handleChange = (_e: ChangeEvent<any>, newValue: string) => {
    if (!newValue.startsWith('#')) {
      setValue(newValue)
    }

    if (onChange) onChange(newValue)
  }

  return (
    <BottomNavigationRoot {...other} value={value} onChange={handleChange}>
      {links.map((link) => (
        <BottomNavigationAction key={link.value || link.label} {...link} />
      ))}
    </BottomNavigationRoot>
  )
}

export { BottomNavigation }
export default BottomNavigation
