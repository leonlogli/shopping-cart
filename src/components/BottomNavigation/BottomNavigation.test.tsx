import React from 'react'
import RestoreIcon from '@material-ui/icons/Restore'
import FavoriteIcon from '@material-ui/icons/Favorite'

import { mountWithTheme } from '../../utils'

import BottomNavigation from './BottomNavigation'

describe('<BottomNavigation />', () => {
  const wrapper = mountWithTheme(
    <BottomNavigation
      links={[
        { label: 'Restaurants', value: '/restaurants', icon: <RestoreIcon /> },
        { label: 'Menu', value: '/menu', icon: <FavoriteIcon /> },
      ]}
    />
  )

  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
