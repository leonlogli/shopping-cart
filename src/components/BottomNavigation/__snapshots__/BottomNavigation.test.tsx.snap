// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomNavigation /> should match a snapshot 1`] = `
<BottomNavigation
  links={
    Array [
      Object {
        "icon": <Memo(RestoreIcon) />,
        "label": "Restaurants",
        "value": "/restaurants",
      },
      Object {
        "icon": <Memo(FavoriteIcon) />,
        "label": "Menu",
        "value": "/menu",
      },
    ]
  }
>
  <Styled(WithStyles(ForwardRef(BottomNavigation)))
    onChange={[Function]}
  >
    <WithStyles(ForwardRef(BottomNavigation))
      className="jTgLGz"
      onChange={[Function]}
    >
      <ForwardRef(BottomNavigation)
        className="jTgLGz"
        classes={
          Object {
            "root": "MuiBottomNavigation-root",
          }
        }
        onChange={[Function]}
      >
        <div
          className="MuiBottomNavigation-root jTgLGz"
        >
          <WithStyles(ForwardRef(BottomNavigationAction))
            icon={<Memo(RestoreIcon) />}
            key=".$/restaurants"
            label="Restaurants"
            onChange={[Function]}
            selected={false}
            showLabel={false}
            value="/restaurants"
          >
            <ForwardRef(BottomNavigationAction)
              classes={
                Object {
                  "iconOnly": "MuiBottomNavigationAction-iconOnly",
                  "label": "MuiBottomNavigationAction-label",
                  "root": "MuiBottomNavigationAction-root",
                  "selected": "Mui-selected",
                  "wrapper": "MuiBottomNavigationAction-wrapper",
                }
              }
              icon={<Memo(RestoreIcon) />}
              label="Restaurants"
              onChange={[Function]}
              selected={false}
              showLabel={false}
              value="/restaurants"
            >
              <WithStyles(ForwardRef(ButtonBase))
                className="MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                focusRipple={true}
                onClick={[Function]}
              >
                <ForwardRef(ButtonBase)
                  className="MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                  classes={
                    Object {
                      "disabled": "Mui-disabled",
                      "focusVisible": "Mui-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  focusRipple={true}
                  onClick={[Function]}
                >
                  <button
                    className="MuiButtonBase-root MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiBottomNavigationAction-wrapper"
                    >
                      <ForwardRef>
                        <WithStyles(ForwardRef(SvgIcon))>
                          <ForwardRef(SvgIcon)
                            classes={
                              Object {
                                "colorAction": "MuiSvgIcon-colorAction",
                                "colorDisabled": "MuiSvgIcon-colorDisabled",
                                "colorError": "MuiSvgIcon-colorError",
                                "colorPrimary": "MuiSvgIcon-colorPrimary",
                                "colorSecondary": "MuiSvgIcon-colorSecondary",
                                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                "root": "MuiSvgIcon-root",
                              }
                            }
                          >
                            <svg
                              aria-hidden={true}
                              className="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
                              />
                            </svg>
                          </ForwardRef(SvgIcon)>
                        </WithStyles(ForwardRef(SvgIcon))>
                      </ForwardRef>
                      <span
                        className="MuiBottomNavigationAction-label MuiBottomNavigationAction-iconOnly"
                      >
                        Restaurants
                      </span>
                    </span>
                    <WithStyles(memo)
                      center={false}
                    >
                      <ForwardRef(TouchRipple)
                        center={false}
                        classes={
                          Object {
                            "child": "MuiTouchRipple-child",
                            "childLeaving": "MuiTouchRipple-childLeaving",
                            "childPulsate": "MuiTouchRipple-childPulsate",
                            "ripple": "MuiTouchRipple-ripple",
                            "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                            "rippleVisible": "MuiTouchRipple-rippleVisible",
                            "root": "MuiTouchRipple-root",
                          }
                        }
                      >
                        <span
                          className="MuiTouchRipple-root"
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            component={null}
                            exit={true}
                          />
                        </span>
                      </ForwardRef(TouchRipple)>
                    </WithStyles(memo)>
                  </button>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(BottomNavigationAction)>
          </WithStyles(ForwardRef(BottomNavigationAction))>
          <WithStyles(ForwardRef(BottomNavigationAction))
            icon={<Memo(FavoriteIcon) />}
            key=".$/menu"
            label="Menu"
            onChange={[Function]}
            selected={false}
            showLabel={false}
            value="/menu"
          >
            <ForwardRef(BottomNavigationAction)
              classes={
                Object {
                  "iconOnly": "MuiBottomNavigationAction-iconOnly",
                  "label": "MuiBottomNavigationAction-label",
                  "root": "MuiBottomNavigationAction-root",
                  "selected": "Mui-selected",
                  "wrapper": "MuiBottomNavigationAction-wrapper",
                }
              }
              icon={<Memo(FavoriteIcon) />}
              label="Menu"
              onChange={[Function]}
              selected={false}
              showLabel={false}
              value="/menu"
            >
              <WithStyles(ForwardRef(ButtonBase))
                className="MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                focusRipple={true}
                onClick={[Function]}
              >
                <ForwardRef(ButtonBase)
                  className="MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                  classes={
                    Object {
                      "disabled": "Mui-disabled",
                      "focusVisible": "Mui-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  focusRipple={true}
                  onClick={[Function]}
                >
                  <button
                    className="MuiButtonBase-root MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiBottomNavigationAction-wrapper"
                    >
                      <ForwardRef>
                        <WithStyles(ForwardRef(SvgIcon))>
                          <ForwardRef(SvgIcon)
                            classes={
                              Object {
                                "colorAction": "MuiSvgIcon-colorAction",
                                "colorDisabled": "MuiSvgIcon-colorDisabled",
                                "colorError": "MuiSvgIcon-colorError",
                                "colorPrimary": "MuiSvgIcon-colorPrimary",
                                "colorSecondary": "MuiSvgIcon-colorSecondary",
                                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                "root": "MuiSvgIcon-root",
                              }
                            }
                          >
                            <svg
                              aria-hidden={true}
                              className="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                              />
                            </svg>
                          </ForwardRef(SvgIcon)>
                        </WithStyles(ForwardRef(SvgIcon))>
                      </ForwardRef>
                      <span
                        className="MuiBottomNavigationAction-label MuiBottomNavigationAction-iconOnly"
                      >
                        Menu
                      </span>
                    </span>
                    <WithStyles(memo)
                      center={false}
                    >
                      <ForwardRef(TouchRipple)
                        center={false}
                        classes={
                          Object {
                            "child": "MuiTouchRipple-child",
                            "childLeaving": "MuiTouchRipple-childLeaving",
                            "childPulsate": "MuiTouchRipple-childPulsate",
                            "ripple": "MuiTouchRipple-ripple",
                            "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                            "rippleVisible": "MuiTouchRipple-rippleVisible",
                            "root": "MuiTouchRipple-root",
                          }
                        }
                      >
                        <span
                          className="MuiTouchRipple-root"
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            component={null}
                            exit={true}
                          />
                        </span>
                      </ForwardRef(TouchRipple)>
                    </WithStyles(memo)>
                  </button>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(BottomNavigationAction)>
          </WithStyles(ForwardRef(BottomNavigationAction))>
        </div>
      </ForwardRef(BottomNavigation)>
    </WithStyles(ForwardRef(BottomNavigation))>
  </Styled(WithStyles(ForwardRef(BottomNavigation)))>
</BottomNavigation>
`;
